src/contexts/application-context/index.tsx

```diff
@@ -0,0 +1,64 @@
+import { createContext, useEffect, useState } from 'react';
+import type { ApplicationContextParams, ApplicationProviderProps } from './types';
+import useCookie from 'react-use-cookie';
+
+export const ApplicationContext = createContext<ApplicationContextParams>({
+  loadingSessionToken: true
+});
+
+export function ApplicationProvider  ({
+  children
+}: ApplicationProviderProps) {
+  const [loadingSessionToken, setLoadingSessionToken] = useState<boolean>(true);
+  const [tokenCookie, setTokenCookie] = useCookie('token', '0');
+  const [temporarySessionToken, setTemporarySessionToken] = useState<string>('');
+
+  useEffect(() => {
+    if (!temporarySessionToken) return setLoadingSessionToken(false);
+
+    setTokenCookie(
+      temporarySessionToken,
+      {
+        days: 365,
+        SameSite: 'Strict',
+        Secure: true
+      }
+    );
+
+    setLoadingSessionToken(false);
+  }, [temporarySessionToken]);
+
+  useEffect(() => {
+    if(tokenCookie !== '0') return;
+
+    getTemporarySessionToken();
+  }, []);
+
+  const getTemporarySessionToken = async () => {
+    const respose = await fetch(
+      `${import.meta.env.VITE_API_URI}/api/v1/temporary-session`,
+      {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({}),
+      }
+    )
+
+    const data = await respose.json();
+    setTemporarySessionToken(data.token);
+  }
```

Comment:
I think this could be cleaner if you take the `setTemporarySessionToken` call out of this function. If you do that, you could take the `getTemporarySessionToken` function out of the context provider definition and it would allow you to write the logic you have split in 2 `useEffect`s above only in a single one. It would also allow become `getTemporarySessionToken` easier to test since it won't depend on React state

```tsx
// somewhere outside the ApplicationProvider
const getTemporarySessionToken = async () => {
  const respose = await fetch(
    `${import.meta.env.VITE_API_URI}/api/v1/temporary-session`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({}),
    }
  )

  const data = await respose.json();
  return data
}


// inside the ApplicationProvider
useEffect(() => {
  (async function() {
    if(tokenCookie !== '0') return;

    const newToken = await getTemporarySessionToken()

    // not sure why you need to handle this scenario
    if (!newToken) {
      setLoadingSessionToken(false);
      return;
    }

    setTokenCookie(
      newToken,
      {
        days: 365,
        SameSite: 'Strict',
        Secure: true
      }
    );

    setLoadingSessionToken(false);
  }())
}, [])
```

What do you think? I'm wrapping it all inside an IIFE because of the `async/await` stuff

Reply:
uff nice, I used the IIFE but the getTemporarySessionToken in the provider